suplemon/suplemon_module.py

```diff
@@ -3,12 +3,111 @@
 Base class for extension modules to inherit.
 """
 
+import os
+import json
 import logging
 
 
+class Storage:
+    """
+    Semi-automatic key/value store for suplemon modules.
+    """
+    def __init__(self, module):
+        self.data = {}
+        self.automatic = False
+        self.module = module
+        self.data_subdir = "modules"
+        self.extension = "json"
+        self.storage_dir = os.path.join(self.module.app.config.get_config_dir(), self.data_subdir)
+        if not os.path.exists(self.storage_dir):
+            try:
+                os.makedirs(self.storage_dir)
+            except:
+                self.app.logger.warning("Module storage folder '{0}' doesn't exist and couldn't be created.".format(
+                                        self.storage_dir))
+
+    def __getitem__(self, i):
+        """Get a storage key value."""
+        return self.data[i]
+
+    def __setitem__(self, i, v):
+        """Set a storage key value."""
+        self.data[i] = v
+        if self.automatic:
+            self.store()
+
+    def __str__(self):
+        """Convert entire data dict to string."""
+        return str(self.data)
+
+    def __len__(self):
+        """Return length of data dict."""
+        return len(self.data)
+
+    def keys(self):
+        return self.data.keys()
+
+    def items(self):
+        return self.data.items()
+
+    def get_path(self):
+        """Get the storage file path."""
+        if self.module.get_name():
+            return os.path.join(self.storage_dir, self.module.get_name() + "." + self.extension)
```

Comment:
Hmm. I should have thought about that a little more. However at the moment the module name is actually ignored, and the filename (without `.py`) is used instead. So it shouldn't be exploitable now. I'll keep this in mind when I get to refactoring the module system.